toolbox/library/src/YarpRpc.cpp

```diff
@@ -0,0 +1,219 @@
+    // No outputs
+    //
+
+
+    const bool ok = blockInfo->setPortsInfo(
+        {
+            // Inputs
+            {InputIndex::TriggerSignal, Port::Dimensions{1}, Port::DataType::DOUBLE},
+        },
+        {
+            // Outputs
+        });
+
+    if (!ok) {
+        bfError << "Failed to configure input / output ports.";
+        return false;
+    }
+
+    return true;
+}
+
+bool YarpRpc::initialize(BlockInformation* blockInfo)
+{
+    if (!Block::initialize(blockInfo)) {
+        return false;
+    }
+
+    using namespace yarp::os;
+
+    // PARAMETERS
+    // ==========
+
+    if (!YarpRpc::parseParameters(blockInfo)) {
+        bfError << "Failed to parse parameters.";
+        return false;
+    }
+
+    // Temp variable to store rpc command string
+    std::string command;
+
+    bool ok = true;
+    ok = ok && m_parameters.getParameter("PortName", pImpl->serverPortName);
+    ok = ok && m_parameters.getParameter("RpcCommand", command);
+
+    if (!ok) {
+        bfError << "Failed to read input parameters.";
+        return false;
+    }
+
+    // CLASS INITIALIZATION
+    // ====================
+
+    pImpl->network = std::make_unique<yarp::os::Network>();
+
+    if (!yarp::os::Network::initialized() || !yarp::os::Network::checkNetwork(5.0)) {
+        bfError << "YARP server wasn't found active.";
+        return false;
+    }
+
+    // Open a rpc client port
+    // TODO: Check if the client name can be improved
+    pImpl->clientPortName = "/rpcClient" + pImpl->serverPortName;
+
+    if (!pImpl->rpcClientPort.open(pImpl->clientPortName)) {
+        bfError << "Error while opening yarp rpc server port.";
+        return false;
+    }
+
+    if (!yarp::os::Network::connect(pImpl->rpcClientPort.getName(), pImpl->serverPortName)) {
+        bfError << "Failed to connect " << pImpl->serverPortName << " to "
+                << pImpl->rpcClientPort.getName() << ".";
+        return false;
+    }
+
+    // Store rpc command in bottle
+    pImpl->rpcCommand.fromString(command);
+
+    return true;
+}
+
+bool YarpRpc::terminate(const BlockInformation* /*blockInfo*/)
+{
+    if(pImpl->rpcClientPort.isWriting()) {
+        bfWarning << "Interrupting rpc port";
+        pImpl->rpcClientPort.interrupt();
+    }
+
+    yarp::os::Network::disconnect(pImpl->rpcClientPort.getName(), pImpl->serverPortName);
+
+    // Close the port
+    pImpl->rpcClientPort.close();
+
+    return true;
+}
+
+bool YarpRpc::output(const BlockInformation* blockInfo)
+{
+    InputSignalPtr triggerSignal = blockInfo->getInputPortSignal(InputIndex::TriggerSignal);
+
+    if (!triggerSignal) {
+        bfError << "Input signal not valid.";
+        return false;
+    }
+
+    yarp::os::Bottle response;
+
+    double trigger = triggerSignal->get<double>(0);
+
+    if (static_cast<bool>(trigger) && !pImpl->rpcClientPort.isWriting()) {
+        pImpl->rpcClientPort.write(pImpl->rpcCommand, response);
+        bfWarning << response.toString().c_str();
+    }
```

Comment:
Is the `isWriting` an actual case that can happen? If all RPC calls are blocking, I think that every previous call to the `output` method of this block should have been already concluded.

If this can happen, probably I would split the two conditions, and if the port `isWriting` the block could return `false` and stop the simulation.

Reply:
Done